kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassifierReferenceImpl.kt

```diff
@@ -1,37 +1,77 @@
 package com.google.devtools.ksp.impl.symbol.kotlin
 
 import com.google.devtools.ksp.IdKeyPair
+import com.google.devtools.ksp.IdKeyTriple
 import com.google.devtools.ksp.KSObjectCache
 import com.google.devtools.ksp.symbol.*
-import org.jetbrains.kotlin.analysis.api.components.buildClassType
-import org.jetbrains.kotlin.analysis.api.types.KtUsualClassType
+import org.jetbrains.kotlin.analysis.api.types.KtClassType
+import org.jetbrains.kotlin.analysis.api.types.KtClassTypeQualifier
+import org.jetbrains.kotlin.analysis.api.types.KtTypeParameterType
 
 class KSClassifierReferenceImpl private constructor(
-    internal val ktType: KtUsualClassType,
+    internal val ktType: KtClassType,
+    internal val index: Int,
     override val parent: KSTypeReference?
 ) : KSClassifierReference {
-    companion object : KSObjectCache<IdKeyPair<KtUsualClassType, KSTypeReference?>, KSClassifierReferenceImpl>() {
-        fun getCached(ktType: KtUsualClassType, parent: KSTypeReference?) =
-            cache.getOrPut(IdKeyPair(ktType, parent)) { KSClassifierReferenceImpl(ktType, parent) }
+    companion object : KSObjectCache<IdKeyTriple<KtClassType, Int, KSTypeReference?>, KSClassifierReferenceImpl>() {
+        fun getCached(ktType: KtClassType, parent: KSTypeReference?, index: Int) =
+            cache.getOrPut(IdKeyTriple(ktType, index, parent)) { KSClassifierReferenceImpl(ktType, index, parent) }
     }
+
+    private val classifierReference: KtClassTypeQualifier
+        get() = ktType.qualifiers[index]
+
     override val qualifier: KSClassifierReference? by lazy {
-        ktType.classId.outerClassId?.let {
-            analyze {
-                buildClassType(it)
-            }
-        }?.let { getCached(it as KtUsualClassType, parent) }
+        if (index == 0) {
+            null
```

Comment:
I'm not sure I understand what this means: for index 0, `null` qualifier. Is it like a `<root>` package? Could you clarify this with an example?

Reply:
Basically it is trying to check if there is more qualifiers, therefore, when the index is already at 0, there can't be more qualifiers before current qualifier.

e.g. for a reference of `A.B` where B is a class defined inside A, index 0 means the reference element part at `A`, and 1 means element at `B`, the test covers it but unfortunately the whole test is blocked by some other unrelated bugs (get top level callables).

Reply:
Sorry, just to make sure one more time. So, in your example, for a reference of `A` then, index 0 returns `null` that indicates this one is the top-most qualifier?

Reply:
index 0 returns `A`, when trying to call `qualifier` on `A`, it triggers this if check and returns `null` to indicate there is no further qualifiers.